invenio_rdm_records/links/models.py

```diff
@@ -0,0 +1,177 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2021 TU Wien.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Secret links for sharing access to records."""
+
+import uuid
+from datetime import date, datetime
+
+from invenio_db import db
+from itsdangerous import BadData
+from sqlalchemy_utils import UUIDType
+
+from .permissions import LinkNeed
+from .serializers import SecretLinkSerializer, TimedSecretLinkSerializer
+from .signals import link_created, link_revoked
+
+SUPPORTED_DIGEST_ALGORITHMS = ("HS256", "HS512")
+
+
+class SecretLink(db.Model):
+    """Secret links for sharing permissions on records."""
+
+    __tablename__ = "rdm_records_secret_links"
+
+    id = db.Column(UUIDType, primary_key=True, default=uuid.uuid4)
+    """Secret link ID."""
+
+    token = db.Column(db.Text, nullable=False)
+    """Secret token for link.
+
+    In contrast to Zenodo-AccessRequests, we don't store encrypted tokens, to
+    enable querying the SecretLink table by tokens and improve performance.
+    In case that the database is leaked, all secret links should be revoked.
+    """
+
+    created = db.Column(
+        db.DateTime, nullable=False, default=datetime.utcnow, index=True
+    )
+    """Creation timestamp."""
+
+    expires_at = db.Column(db.DateTime, nullable=True)
+    """Expiration date."""
+
+    permission_level = db.Column(db.String, nullable=False, default="")
+    """Permission level of the link."""
+
+    def covers(self, permission):
```

Comment:
Looking at tests, would go for `permission` -> `action`

(and could also go for `covers` -> `allows`)

but maybe the idea is to change this to an actual permission level ("Metadata curator" / "Owner" ...) down the road?